modules/assim.sequential/R/sda.enkf.R

```diff
@@ -240,42 +255,58 @@ sda.enkf <- function(settings, IC, prior, obs.mean, obs.cov, variables,
     
     obs = !is.na(obs.mean[[t]])
     
-    ### ANALYSIS
-    if(any(obs)){ #if no obs skip analysis
+    ###-------------------------------------------------------------------###
+    ### analysis                                                          ###
+    ###-------------------------------------------------------------------###  
+    if(any(obs)){ #if no observations skip analysis
     mu.f = as.numeric(apply(X,2,mean,na.rm=TRUE))
     Pf   = cov(X)
-    Y    = obs.mean[[t]][[1]][pmatch(colnames(X), names(obs.mean[[t]][[1]]))]
+    Y    = obs.mean[[t]][pmatch(colnames(X), names(obs.mean[[t]]))]
 
-    H = diag(length(obs.mean[[t]][[1]]))
-    R = obs.cov[[t]]#diag(as.numeric(obs.sd[[t]][[1]][pmatch(colnames(X),names(obs.mean[[t]][[1]]))])^2)
+    R = as.matrix(obs.cov[[t]])#how do we make sure this is in the right order?
     
-    for(s in 1:length(obs.mean[[t]][[1]])){
-      if(diag(R)[s]==0){
-        diag(R)[s] <- .01^2 #probably way low?
+    if(length(obs.mean[[t]])>1){
+      for(s in 1:length(obs.mean[[t]][[1]])){
+        if(diag(R)[s]==0){ #if covariance is 0 then set it so model with
+          diag(R)[s] <- min(diag(R)[which(diag(R)!=0)])/2#.01^2 #probably way low?
+        }
       }
     }
-   
+
+    ###-------------------------------------------------------------------###
+    ### Kalman Filter                                                     ###
+    ###-------------------------------------------------------------------###
     if(processvar == FALSE){
+      H = diag(ncol(X)) #correct fix?
+      
       K    = Pf%*%t(H)%*%solve(R+H%*%Pf%*%t(H))
       mu.a = mu.f + K%*%(Y-H%*%mu.f)
       Pa   = (diag(ncol(X)) - K%*%H)%*%Pf
     } else { 
       
+      ###-------------------------------------------------------------------###
+      ### Generalized Ensemble Filter                                       ###
+      ###-------------------------------------------------------------------###
+      
       #### initial conditions
       bqq[1] <- length(mu.f)
       aqq[1,,] <- diag(length(mu.f))*bqq[1]
       
-      #### extinct in all ensemble vector
-      E <- which(colSums(X)==0)
+      #### extinct from the forecast ensembles
+      E <- which(colSums(Pf) == 0)
       
-      #### forcast to model vector
+      #### forcast to modeled vector ## remove spp. that don't show up in forecast
       F2M<- seq(1,length(mu.f),1)
       F2M[E]<-NA
       F2M <- na.omit(F2M)
       
-      #### translating x to y so remove extinct species
+      #### translating x to y ## remove X that you don't have data for
       X2Y<- seq(1,length(Y),1)
-      X2Y <- X2Y[!is.na(Y)]
+      X2Y <- X2Y[which(!is.na(Y))]
+      
+      for(i in 1:length(diag(Pf))){
+        if(diag(Pf)[i]<.0000001) diag(Pf)[i]<-1 #HACK
```

Comment:
Oh I see. Well, the reason it's a problem is because solve won't work because it's base tolerance is 2.220446e-16. You can set it to 0 but then JAGS gets upset. So, maybe I'll just set it to the default tolerance for solve?